db/db_impl_compaction_flush.cc

```diff
@@ -1541,27 +1542,49 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
       c.reset(cfd->PickCompaction(*mutable_cf_options, log_buffer));
       TEST_SYNC_POINT("DBImpl::BackgroundCompaction():AfterPickCompaction");
       if (c != nullptr) {
-        // update statistics
-        MeasureTime(stats_, NUM_FILES_IN_SINGLE_COMPACTION,
-                    c->inputs(0)->size());
-        // There are three things that can change compaction score:
-        // 1) When flush or compaction finish. This case is covered by
-        // InstallSuperVersionAndScheduleWork
-        // 2) When MutableCFOptions changes. This case is also covered by
-        // InstallSuperVersionAndScheduleWork, because this is when the new
-        // options take effect.
-        // 3) When we Pick a new compaction, we "remove" those files being
-        // compacted from the calculation, which then influences compaction
-        // score. Here we check if we need the new compaction even without the
-        // files that are currently being compacted. If we need another
-        // compaction, we might be able to execute it in parallel, so we add it
-        // to the queue and schedule a new thread.
-        if (cfd->NeedsCompaction()) {
-          // Yes, we need more compactions!
+        uint64_t size_added_by_compaction = 0;
+        // First check if we even have the space to do the compaction
+        for (size_t i = 0; i < c->num_input_levels(); i++) {
+          for (size_t j = 0; j < c->num_input_files(i); j++) {
+            FileMetaData *filemeta = c->input(i, j);
+            size_added_by_compaction += filemeta->compensated_file_size;
+          }
+        }
+        // Now figure out how much free space is in the file filesystem
+        struct statvfs statbuf;
+        statvfs(dbname_.c_str(), &statbuf);
```

Comment:
there's a couple issues we need to think more about:

1. this function doesn't exist in Windows, which looks like the cause of travis failing. Maybe make it a function on Env, make it return `Status::NotSupported` by default, and override with a call to `statvfs` in PosixEnv.
2. `dbname_` isn't always the directory compactions will run in. There's also `db_paths` which can spread SST files across multiple directories. I'm ok with just skip this logic if people are setting multiple directories, which is rare anyways.

Reply:
For 2., the compaction affects the directory to which it outputs files, which may not be `dbname_`.